README.md
# esc - Effect Size Computation for Meta Analysis <img src="man/figures/logo.png" align="right" />

This is an R implementation of the web-based 'Practical Meta-Analysis Effect Size Calculator' from David B. Wilson. The original calculator can be found at http://www.campbellcollaboration.org/escalc/html/EffectSizeCalculator-Home.php.

Based on the input, the effect size can be returned as standardized mean difference (`d`), Hedges' `g`, correlation coefficient effect size `r` or Fisher's transformation `z`, odds ratio or log odds effect size.

### Return values

The return value of all functions has the same structure:

* The effect size, whether being `d`, `g`, `r`, (Cox) odds ratios or (Cox) logits, is always named `es`.
* The standard error of the effect size, `se`.
* The variance of the effect size, `var`.
* The lower and upper confidence limits `ci.lo` and `ci.hi`.
* The weight factor, based on the inverse-variance, `w`.
* The total sample size `totaln`.
* The effect size measure, `measure`, which is typically specified via the `es.type`-argument.
* Information on the effect-size conversion, `info`.
* A string with the study name, if the `study`-argument was specified in function calls.

#### Correlation Effect Size

If the correlation effect size `r` is computed, the transformed Fisher's z and their confidence intervals are also returned. The variance and standard error for the correlation effect size r are always based on Fisher's transformation.

#### Odds Ratio Effect Size

For odds ratios, the variance and standard error are always returned on the log-scale!

### S3 methods

The `esc` package offers the S3 methods `print` and `as.data.frame`

### Combining results into a single data frame

The `combine_esc` method is a convenient way to create pooled data frames of different effect size calculations, for further use. Here is an example of `combine_esc`, which returns a `data.frame` object.

```
e1 <- esc_2x2(grp1yes = 30, grp1no = 50, grp2yes = 40, grp2no = 45, study = "Study 1")
e2 <- esc_2x2(grp1yes = 30, grp1no = 50, grp2yes = 40, grp2no = 45, es.type = "or", study = "Study 2")
e3 <- esc_t(p = 0.03, grp1n = 100, grp2n = 150, study = "Study 3")
e4 <- esc_mean_sd(grp1m = 7, grp1sd = 2, grp1n = 50, grp2m = 9, 
                  grp2sd = 3, grp2n = 60, es.type = "logit", study = "Study 4")

combine_esc(e1, e2, e3, e4)

>     study         es    weight sample.size        se        var       ci.lo      ci.hi measure
> 1 Study 1 -0.3930426  9.944751         165 0.3171050 0.10055556 -1.01455689  0.2284717   logit
> 2 Study 2  0.6750000  9.944751         165 0.3171050 0.10055556  0.36256305  1.2566780      or
> 3 Study 3  0.2817789 59.433720         250 0.1297130 0.01682547  0.02754605  0.5360117       d
> 4 Study 4 -1.3981827  7.721145         110 0.3598812 0.12951447 -2.10353685 -0.6928285   logit
```


**esc** is still under development, i.e. not all effect size computation options are implemented yet. The remaining options will follow in further updates.

## Installation

### Latest development build

To install the latest development snapshot (see latest changes below), type following commands into the R console:

```r
library(devtools)
devtools::install_github("strengejacke/esc")
```

### Officiale, stable release

[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/esc)](https://cran.r-project.org/package=esc)
&#160;&#160;
[![downloads](http://cranlogs.r-pkg.org/badges/esc)](http://cranlogs.r-pkg.org/)
&#160;&#160;
[![total](http://cranlogs.r-pkg.org/badges/grand-total/esc)](http://cranlogs.r-pkg.org/)

To install the latest stable release from CRAN, type following command into the R console:

```r
install.packages("esc")
```

## Citation

In case you want / have to cite my package, please use `citation('esc')` for citation information.


NEWS
Version 0.3.2
------------------------------------------------------------------------------
Bug fixes:
* `combine_esc()` did not work for effect sizes computed with `esc_t()`, when the returned effect size was `r`.

Version 0.3.1
------------------------------------------------------------------------------
Bug fixes:
* `effect_sizes()` did not always properly extract study names and used numeric indices instead character values.

Version 0.3.0
------------------------------------------------------------------------------
New functions:
* `effect_sizes()` to generate a data frame with results of multiple effect size computations, based on data from another data frame.

Bug fixes:
* Fix control statements with condition with greater than one, which currently give a warning, but in future R versions may throw an error.


Version 0.2.0
------------------------------------------------------------------------------
New functions:
* `write_esc()` as a convenient wrapper to write results to an Excel csv-file.

Version 0.1.0
------------------------------------------------------------------------------
General:
* Initial release on CRAN.

R/esc_2x2.R
#' @title Compute effect size from 2 by 2 Contingency Table
#' @name esc_2x2
#'
#' @description Compute effect size from a 2 by 2 frequency table.
#'
#' @param grp1yes Size of treatment group with successes (outcome = yes).
#' @param grp1no Size of treatment group with non-successes (outcome = no).
#' @param grp2yes Size of control group with successes (outcome = yes).
#' @param grp2no Size of control group with non-successes (outcome = no).
#'
#' @inheritParams esc_beta
#' @inheritParams esc_t
#'
#' @return The effect size \code{es}, the standard error \code{se}, the variance
#'         of the effect size \code{var}, the lower and upper confidence limits
#'         \code{ci.lo} and \code{ci.hi}, the weight factor \code{w} and the
#'         total sample size \code{totaln}.
#'
#' @note If \code{es.type = "r"}, Fisher's transformation for the effect size
#'       \code{r} and their confidence intervals are also returned.
#'
#' @references Lipsey MW, Wilson DB. 2001. Practical meta-analysis. Thousand Oaks, Calif: Sage Publications
#'             \cr \cr
#'             Wilson DB. 2016. Formulas Used by the "Practical Meta-Analysis Effect Size Calculator". Unpublished manuscript: George Mason University
#'
#' @examples
#' # effect size log odds
#' esc_2x2(grp1yes = 30, grp1no = 50, grp2yes = 40, grp2no = 45)
#'
#' # effect size odds ratio
#' esc_2x2(grp1yes = 30, grp1no = 50, grp2yes = 40, grp2no = 45, es.type = "or")
#'
#' @export
esc_2x2 <- function(grp1yes,
                    grp1no,
                    grp2yes,
                    grp2no,
                    es.type = c("logit", "d", "g", "or", "r", "cox.d"),
                    study = NULL,
                    ...) {
  # match  arguments
  es.type <- match.arg(es.type)


  # evaluate ellipses
  ell <- match.call(expand.dots = FALSE)$`...`
  # do we have info string?
  if (!is.null(ell$info))
    info <- ell$info
  else
    # if not, set default
    info <- "2x2 table (OR)"


  es <- (grp1yes * grp2no) / (grp1no * grp2yes)
  v <- 1 / grp1yes + 1 / grp1no + 1 / grp2yes + 1 / grp2no

  totaln <- grp1yes + grp1no + grp2yes + grp2no

  # which es type to be returned?
  if (es.type == "or") {
    return(structure(
      class = c("esc", "esc_2x2"),
      list(es = es, se = sqrt(v), var = v, ci.lo = exp(lower_d(log(es), v)),
           ci.hi = exp(upper_d(log(es), v)), w = 1 / v, totaln = totaln,
           measure = "or", info = paste0(info, " coefficient to effect size odds ratio"),
           study = study)
    ))
  }

  # which es type to be returned?
  if (es.type == "d") return(esc_or2d(or = es, v = v, totaln = totaln, es.type = "d", info = paste0(info, " to effect size d"), study = study))

  # which es type to be returned?
  if (es.type == "cox.d") return(esc_or2d(or = es, v = v, totaln = totaln, es.type = "cox.d", info = paste0(info, " to effect size Cox d"), study = study))

  # which es type to be returned?
  if (es.type == "g") return(esc_or2d(or = es, v = v, totaln = totaln, es.type = "g", info = paste0(info, " to effect size Hedges' g"), study = study))

  # convert to plain d
  es <- log(es) * sqrt(3) / pi
  v <- v * 3 / pi ^ 2

  # which es type to be returned?
  if (es.type == "logit") return(esc_d2logit(d = es, v = v, totaln = totaln, info = paste0(info, " to effect size logits"), study = study))

  # which es type to be returned?
  if (es.type == "r") return(esc_d2r(d = es, v = v, grp1n = (grp1yes + grp1no), grp2n = (grp2yes + grp2no), info = paste0(info, " coefficient to effect size correlation"), study = study))
}
